openedx/core/apidocs.py

```diff
@@ -0,0 +1,161 @@
+            text = rest
+    return textwrap.dedent(text)
+
+
+def schema(
+    manual_parameters=None,
+    operation_summary=None,
+    operation_description=None
+):
+    """
+    Decorator for documenting an API endpoint.
+
+    This is heavily inspired from the the `drf_yasg.utils.swagger_auto_schema`__
+    decorator, but callers do not need to know about this abstraction.
+    All description fields should be in Markdown.
+
+    __ https://drf-yasg.readthedocs.io/en/stable/drf_yasg.html#drf_yasg.utils.swagger_auto_schema
+    """
+    for param in manual_parameters or ():
+        param.description = dedent(param.description)
+
+    def decorator(view_func):
+        """
+        Final view decorator.
+        """
+        if view_func.__doc__ is not None:
+            doc_lines = view_func.__doc__.strip().split("\n")
+            if operation_summary is None and doc_lines:
+                operation_summary = doc_lines[0].strip()
+            if operation_description is None and len(doc_lines) > 1:
+                operation_description = "\n".join(doc_lines[1:])
+        if operation_description is not None:
+            operation_description = dedent(operation_description)
+        return drf_swagger_auto_schema(
+            manual_parameters=manual_parameters,
+            operation_summary=operation_summary,
+            operation_description=operation_description
+        )(view_func)
+    return decorator
+
+
+def is_schema_request(request):
+    """Is this request serving an OpenAPI schema?"""
+    return request.query_params.get('format') == 'openapi'
+
+
+class ParameterLocation(object):
+    """Location of API parameter in request."""
+    BODY = openapi.IN_BODY
+    PATH = openapi.IN_PATH
+    QUERY = openapi.IN_QUERY
+    FORM = openapi.IN_FORM
+    HEADER = openapi.IN_HEADER
+
+
+def string_parameter(name, in_, description=None):
+    """
+    Convenient function for defining a string parameter.
+
+    Args:
+        name (str)
+        in_ (ParameterLocation attribute)
+        description (str)
+    """
+    return parameter(name, in_, type=str, description=description)
+
+
+def parameter(name, in_, type=None, description=None):
+    """
+    Define typed parameters.
+
+    Args:
+        name (str)
+        in_ (ParameterLocation attribute)
+        type (type): one of object, str, float, int, bool, list, file.
+        description (str)
+    """
+    openapi_type = None
+    if type is object:
+        openapi_type = openapi.TYPE_OBJECT
+    elif type is str:
+        openapi_type = openapi.TYPE_STRING
+    elif type is float:
+        openapi_type = openapi.TYPE_NUMBER
+    elif type is int:
+        openapi_type = openapi.TYPE_INTEGER
+    elif type is bool:
+        openapi_type = openapi.TYPE_BOOLEAN
+    elif type is list:
+        openapi_type = openapi.TYPE_ARRAY
+    elif type is file:
+        openapi_type = openapi.TYPE_FILE
+
+    else:
+        raise RuntimeError("Unsupported parameter type")
```

Comment:
I think `ValueError` is appropriate here. Also, including the value of the bad parameter in the message would be helpful for debugging.

Reply:
:+1: